### PR TITLE
[Flink][Improvement]Throw an exception when the table has no primary key set instead of catching it

### PR DESCRIPTION
**Pull request by JasonLeeCoding**
Throw an exception when the table has no primary key set instead of catching it,In this way, the flink job will failover, and users can find the problem faster.

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 1]
```diff
@@ -275,9 +275,9 @@ private void splitTable(TableId nextTable) {
 
     @Override
     public Optional<MySqlSplit> getNext() {
-        checkSplitterErrors();
         waitTableDiscoveryReady();
         synchronized (lock) {
+            checkSplitterErrors();
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:280]: We should also make the assignment logic of the `uncaughtSplitterException` guarded by the `notificationLock` in the method `splitChunksForRemainingTables`.

**Review by ruanhang1993 — COMMENTED** [review]
@JasonLeeCoding Thanks for contributing. I left some comments.
Would you mind change the commit message to '[mysql] Fix race condition between split and coordinator thread that async exception might not be caught'?

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java (hunk 1 above)
Comment (JasonLeeCoding) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:280]: hi @ruanhang1993 I've already modified it,Please help to review it when you have time,thanks.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:1]: Thanks for your quick reply, @JasonLeeCoding .
Actually we do not need to put all parts of `splitChunksForRemainingTables` under the lock.
We could modify like this.
```java
private void splitChunksForRemainingTables() {
    try {
        ......
    } catch (Throwable e) {
        // Release the potential waiting getNext() call
        synchronized (lock) {
            if (uncaughtSplitterException == null) {
                uncaughtSplitterException = e;
            } else {
                uncaughtSplitterException.addSuppressed(e);
            }
            lock.notify();
        }
    }
}
```

**Review by ruanhang1993 — APPROVED** [review]
LGTM. Thanks~

**Pull request merged**